internal/adaptorsigs/adaptor.go

```diff
@@ -0,0 +1,489 @@
+package adaptorsigs
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/decred/dcrd/crypto/blake256"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4/schnorr"
+)
+
+// AdaptorSignatureSize is the size of an encoded adaptor Schnorr signature.
+const AdaptorSignatureSize = 129
+
+// scalarSize is the size of an encoded big endian scalar.
+const scalarSize = 32
+
+var (
+	// rfc6979ExtraDataV0 is the extra data to feed to RFC6979 when generating
+	// the deterministic nonce for the EC-Schnorr-DCRv0 scheme.  This ensures
+	// the same nonce is not generated for the same message and key as for other
+	// signing algorithms such as ECDSA.
+	//
+	// It is equal to BLAKE-256([]byte("EC-Schnorr-DCRv0")).
+	rfc6979ExtraDataV0 = [32]byte{
+		0x0b, 0x75, 0xf9, 0x7b, 0x60, 0xe8, 0xa5, 0x76,
+		0x28, 0x76, 0xc0, 0x04, 0x82, 0x9e, 0xe9, 0xb9,
+		0x26, 0xfa, 0x6f, 0x0d, 0x2e, 0xea, 0xec, 0x3a,
+		0x4f, 0xd1, 0x44, 0x6a, 0x76, 0x83, 0x31, 0xcb,
+	}
+)
+
+// AdaptorSignature is a signature with auxillary data that commits to a hidden
+// value. When an adaptor signature is combined with a corresponding signature,
+// the hidden value is revealed. Alternatively, when combined with a hidden
+// value, the adaptor reveals the signature.
+//
+// An adaptor signature is created by either doing a public or private key
+// tweak of a valid schnorr signature. A private key tweak can only be done by
+// a party who knows the hidden value, and a public key tweak can be done by
+// a party that only knows the point on the secp256k1 curve derived by the
+// multiplying the hidden value by the generator point.
+//
+// Generally the workflow of using adaptor signatures is the following:
+//  1. Party A randomly selects a hidden value and creates a private key
+//     modified adaptor signature of something for which party B requires
+//     a valid signature.
+//  2. The Party B sees the PublicTweak in the adaptor signature, and creates
+//     a public key tweaked adaptor signature for something that party A
+//     requires a valid signature.
+//  3. Since party A knows the hidden value, they can use the hidden value to
+//     create a valid signature from the public key tweaked adaptor signature.
+//  4. When the valid signature is revealed, by being posted to the blockchain,
+//     party B can recover the tweak and use it to decrypt the private key
+//     tweaked adaptor signature that party A originally sent them.
+type AdaptorSignature struct {
+	r secp256k1.FieldVal
+	s secp256k1.ModNScalar
+	// t will always be in affine coordinates.
+	t           secp256k1.JacobianPoint
```

Comment:
Taking a look at this again, I think it would be a little harder to make a mistake regarding affine vs Jacobian as wel as being more efficient (one less field val to store) to simply keep the affine point here.  Namely `tX secp256k1.FieldVal` and `tY secp256k1.FieldVal`.

Then, converting it to a `secp256k1.JacobianPoint` in the two places that need it as one then is merely:

```Go
var tJ JacobianPoint
tJ.X.Set(&sig.tX)
tJ.X.Set(&sig.tY)
tJ.Z.SetInt(1)
``` 

If you're concerned about the slight repetition in the two places, you could also do something like:

```Go
type affinePoint struct {
	X secp256k1.FieldVal
	Y secp256k1.FieldVal
}

func (p *affinePoint) AsJacobian() secp256k1.JacobianPoint {
	var result secp256k1.JacobianPoint
	result.X.Set(&p.x)
	result.Y.Set(&p.y)
	result.Z.SetInt(1)
	return result
}

...

type AdaptorSignature struct {
	r secp256k1.FieldVal
	s secp256k1.ModNScalar
	t affinePoint
	pubKeyTweak bool
}
```

That would get inlined and thus not be any more expensive.

Reply:
Nice.